packages/lib-utils/src/app/redux/reducers/k8s/index.ts

```diff
@@ -1,15 +1,271 @@
-import type { ActionType as Action } from 'typesafe-actions';
-import type { K8sModelCommon } from '../../../../types/k8s';
+import * as _ from 'lodash-es';
+import { getReferenceForModel, getNamespacedResources, allModels } from '../../../../k8s/k8s-utils';
+import type { K8sModelCommon, K8sResourceCommon } from '../../../../types/k8s';
 import type { K8sState } from '../../../../types/redux';
+import type { K8sAction } from '../../actions/k8s';
+import { ActionType } from '../../actions/k8s';
+import { getK8sDataById } from './selector';
 
-// TODO remove when actions/reducers are migrated
-const emptyK8sState: K8sState = fromJS({
-  RESOURCES: { inFlight: false, models: ImmutableMap<string, K8sModelCommon>() },
-});
+const getQN: (obj: K8sResourceCommon) => string = (obj: K8sResourceCommon) => {
+  const { name, namespace } = obj?.metadata || {};
+  // Name + namespace is not unique for PackageManifest resources, so include the catalog source.
+  // TODO: We should be able to remove this when the upstream OLM bug is fixed:
+  //       https://bugzilla.redhat.com/show_bug.cgi?id=1814822
+  if (obj.apiVersion === 'packages.operators.coreos.com/v1' && obj.kind === 'PackageManifest') {
+    return `(${obj.status?.catalogSource})-${name}`;
+  }
+  return (namespace ? `(${namespace})-` : '') + name;
+};
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const moreRecent = (a: any, b: any) => {
+  const metaA = a.get('metadata').toJSON();
+  const metaB = b.get('metadata').toJSON();
+  if (metaA.uid !== metaB.uid) {
+    return new Date(metaA.creationTimestamp) > new Date(metaB.creationTimestamp);
+  }
+  return parseInt(metaA.resourceVersion, 10) > parseInt(metaB.resourceVersion, 10);
+};
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const removeFromList = (list: ImmutableMap<string, any>, resource: K8sResourceCommon) => {
+  const qualifiedName = getQN(resource);
+  consoleLogger.info(`deleting ${qualifiedName}`);
+  return list.delete(qualifiedName);
+};
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const updateList = (list: ImmutableMap<string, any>, nextJS: K8sResourceCommon) => {
+  const qualifiedName = getQN(nextJS);
+  const current = list.get(qualifiedName);
+  const next = fromJS(nextJS);
+
+  if (!current) {
+    return list.set(qualifiedName, next);
+  }
+
+  if (!moreRecent(next, current)) {
+    return list;
+  }
+
+  // TODO: (kans) only store the data for things we display ...
+  //  and then only do this comparison for the same stuff!
+  if (
+    current
+      .deleteIn(['metadata', 'resourceVersion'])
+      .equals(next.deleteIn(['metadata', 'resourceVersion']))
+  ) {
+    // If the only thing that differs is resource version, don't fire an update.
+    return list;
+  }
+
+  return list.set(qualifiedName, next);
+};
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const loadList = (oldList: any, resources: any) => {
```

Comment:
oldList is probably the ImmutableMap of ImmutableMaps?

Resources is typed on line 72, so I think we can safely do an array here.

```suggestion
const loadList = (oldList: any, resources: K8sResourceCommon[]) => {
```

Reply:
The `Loaded` action case below was not happy when I tried this :( 
<img width="1303" alt="image" src="https://user-images.githubusercontent.com/21317056/157079256-dd685224-7cea-4725-9ac6-82f2c9c81f58.png">

Reply:
Yeah, I think LoadList has to be probably wrong that it can take a single item 🤔